src/core/execute.c

```diff
@@ -2161,10 +2161,18 @@ static int setup_exec_directory(
                         if (r < 0 && r != -EEXIST)
                                 goto fail;
                         if (r == -EEXIST) {
-                                if (chmod(p, context->directories[type].mode) < 0) {
+                                struct stat st;
+                                mode_t mask = 07777;
+                                r = stat(p, &st);
+                                if (r < 0) {
                                         r = -errno;
                                         goto fail;
                                 }
+                                if ((st.st_mode & mask) != context->directories[type].mode) {
```

Comment:
not that it matters, but i think it would be nicer to also mask the right-hand side of the comparison with 07777. I mean, the mode stored there shouldn't have any other bits set, but I think it would be more readable. i.e. maybe rewrite this check as:

```c
if ((st.st_mode ^ contrext->directories[type].mode) & 07777 != 0))
…
```